### ISSUE TITLE
Intermittent file read error in R Markdown documents

### ISSUE THREAD

**Issue by melissa-barca** [rstudio/rstudio#9284]
### System details

    RStudio Edition : Any
    RStudio Version : 1.4+
    OS Version      : macos 11.2.3 (but likely any)
    R Version       : 4.0.3

### Steps to reproduce the problem
1. Turn on "Save documents when editor loses input focus"  via the command palette.
2. Open a new R Markdown document without a title.
3. Add a chunk with the following code:
```
{r fig.height=2.5}
plot(iris$Sepal.Length, iris$Sepal.Width, main = "Plot 1")
```
4. Run the chunk
5. Continue changing the plot and rerunning it (e.g. Plot 1 -> Plot 2 -> Plot 3 -> Plot 4 -> Plot 1) until you see an error  like either of the following:
```
Error in plot.new() : 
  could not open file '/cloud/project/.Rproj.user/shared/notebooks/31FBD4C0-save-before-switching-mode/1/9439FB1Eb47ff7a1/clmph3wfvzvor_t/_rs_chunk_plot_001.png'
```
```
ERROR 23 Apr 2021 13:07:23 [rsession-rstudio-user] ERROR system error 2 (No such file or directory) [path: /cloud/project/.Rproj.user/shared/notebooks/ADBEFFC5-plot-error-reprex/1/6B852F2486e992d0/cza06q1wj83ws/000009.error]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForWrite(std::shared_ptr<std::basic_ostream<char> >&, bool) const src/cpp/shared_core/FilePath.cpp:1482; LOGGED FROM: void rstudio::session::modules::rmarkdown::notebook::ChunkExecContext::onError(const rstudio::core::json::Object&) src/cpp/session/modules/rmarkdown/NotebookExec.cpp:383|||
```

### Describe the problem in detail
This was brought our attention in support ticket #60902. 

The png file is not in the expected directory but is in a similar `/s/` directory. It seems there is a varying race condition where we are autosaving the file at the same time we are executing the code so the file changes from an unsaved state to a saved state mid-execution. The `/s/` directory is used for the saved state.
This has only been reproducible in Visual Editing mode because the save process is more involved and takes longer.

Originally we were only able to reproduce this in rstudio.cloud because they have configured "Save documents when editor loses input focus" to true when its default is off. The workaround is to turn this setting off.

### Describe the behavior you expected
We should not attempt to save a file while executing chunk code. 

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [x] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by kippandrew** [comment]
Support ticket #62587 also seeing a slight variation of this same issue.

**Comment by kippandrew** [comment]
RStudio Cloud uses the `auto_save_on_blur` option to automatically save documents when the editor loses focus. We're considering disabling that option, and using `auto_save_idle_ms` instead. My question is will that allow us to continue to auto-save documents, but avoid this particular bug?

**Comment by jmcphers** [comment]
The bug happens when the autosave code and the notebook chunk execution code run simultaneously. Switching to idle_ms should get rid of the problem 99.9% of the time but there's still _some_ chance of concurrency (for example, if you set idle_ms to 1000 and then execute a chunk _exactly_ 1 second after you stop typing).

**Comment by jmcphers** [comment]
How long did you wait between steps 3 and 4? If it was less than the autosave interval, this is expected -- the session didn't have time to do an autosave before it crashed. The "blur" setting doesn't have a timer; a blur happens between steps 3 and 4 when you switched to the R console to run the code.

If it was more than the autosave interval, it's a bug.

**Comment by omar-rs** [comment]
> How long did you wait between steps 3 and 4? If it was less than the autosave interval, this is expected -- the session didn't have time to do an autosave before it crashed. The "blur" setting doesn't have a timer; a blur happens between steps 3 and 4 when you switched to the R console to run the code.
> 
> If it was more than the autosave interval, it's a bug.

The time between steps 3 and 4 is definitely more than 1s (the autosave interval).  You can see this in full action in this video:

https://user-images.githubusercontent.com/55405342/129228620-f736c375-7a71-474d-b68d-480ae35eea5c.mov

**Comment by ncalies** [comment]
Potential customer ticket - #65277

**Comment by jmcphers** [comment]
Moving back to PT for consideration. See: https://joshuamrosenberg.com/post/2021/09/14/use-rstudio-cloud-for-teaching-with-caution/

**Comment by jgutman** [comment]
This appears to have been fixed in Prairie Trillium but was never tested. Do we test now, or just close it out?

**Comment by mikebessuille** [comment]
A quick test is warranted: Jonathan confirms this was fixed in PT and never closed.

**Comment by jgutman** [comment]
Did some light testing, and confirmed this appears to be fixed in RStudio Desktop 2022.2.2.485 on Mac OS, and we haven't had any further reports since the fix was introduced. 

Note that the bug was more noticeable in RStudio.cloud than on Desktop but was previously possible on either; it should now be fixed on all platforms

**Comment by samshott** [comment]
I believe I am getting this issue - I haven't seen it documented anywhere else.

**Error**
`Warning in gzfile(file, "wb") :
  cannot open compressed file '/Users/samuelericksen/.local/share/rstudio/notebooks/3672E397-TimeSeriesLecture/1/D7355B516fee4ab7/c9klbc22qpzt6_t/453b1ed4fcff4501970199fb94aabd1a.snapshot', probable reason 'No such file or directory'
Error in gzfile(file, "wb") : cannot open the connection
Error in dev.off() : 
  QuartzBitmap_Output - unable to open file '/Users/samuelericksen/.local/share/rstudio/notebooks/3672E397-TimeSeriesLecture/1/D7355B516fee4ab7/c9klbc22qpzt6_t/_rs_chunk_plot_001.png'`

**System**
Mac with an M1 Max chip. MacOS 14.0 (although it was also occurring previous versions)
Most recent Rstudio build
R version 4.2.1

**Behavior**
The issue occurs intermittently in Visual Code mode in both Quarto and Rmd documents. Going to source mode stops it from occurring. It does seem to be related to the amount of time between editing the code chunk and running it (i.e. if I edit it then run it immediately I get the error). Changing the Idle ms for autosave, and even turning it off doesn't seem to change the behavior.

Interestingly, about half the time this occurs I have to restart my R session, as it wont allow me to run anything else otherwise. Also, Changing my graphics backend from Quartz to Cairo gets rid of the second part of the error, but all other behavior remains the same.
